File: Lecture 3/Python/gauss_seidel.py
from create_Yb import create_Yb

import numpy as np
import cmath
import pickle


def gauss_seidel(system):
    """
    Perform power flow analysis using Gauss-Seidel's power flow method.
    """

    # Extract the system data
    number_of_buses = system['buses'].shape[0]
    bus_type = system['buses'][:, 1].astype(int)
    v_specified = system['buses'][:, 2]
    theta = system['buses'][:, 3]
    p_load = system['buses'][:, 4]
    q_load = system['buses'][:, 5]
    p_gen = system['buses'][:, 6]
    q_gen = system['buses'][:, 7]
    q_gen_min = system['buses'][:, 8]
    q_gen_max = system['buses'][:, 9]

    # Create the bus-admittance matrix
    Y_b = create_Yb(system)

    # Initialize the calculation variables
    p = p_gen - p_load
    q = q_gen - q_load
    v = v_specified * np.exp(1j * np.radians(theta))
    v_old = np.zeros_like(v)
    iteration = 0
    tolerance = 1e-6

    # Precompute reciprocal of diagonal elements of Y_b
    Y_b_diag_inv = 1.0 / np.diag(Y_b)

    # Main loop
    while np.any(np.abs(v - v_old) > tolerance):
        # Update the iteration variables
        v_old = np.copy(v)
        iteration += 1

        # Apply the Gauss-Seidel power flow equations to each node except the slack node
        for i in range(1, number_of_buses):
            # Calculate the sum of node current contributions
            S = np.dot(Y_b[i, :], v)
            S -= Y_b[i, i] * v[i]

            if bus_type[i] == 2:
                # PV node
                q[i] = -np.imag(np.conj(v[i]) * (S + Y_b[i, i] * v[i]))
                q_gen[i] = q[i] + q_load[i]

                # Check the generator reactive power limits
                if q_gen[i] < q_gen_min[i] or q_gen[i] > q_gen_max[i]:
                    q_gen[i] = max(min(q_gen[i], q_gen_max[i]), q_gen_min[i])
                    q[i] = q_gen[i] - q_load[i]
                    v[i] = Y_b_diag_inv[i] * ((p[i] - 1j * q[i]) / np.conj(v[i]) - S)
                else:
                    v[i] = Y_b_diag_inv[i] * ((p[i] - 1j * q[i]) / np.conj(v[i]) - S)
                    v[i] = v_specified[i] * cmath.exp(1j * cmath.phase(v[i]))

            else:
                # PQ node
                v[i] = Y_b_diag_inv[i] * ((p[i] - 1j * q[i]) / np.conj(v[i]) - S)

    # Evaluate the voltage magnitudes and phase angles
    theta = np.angle(v, deg=True)
    v = np.abs(v)

    return v, theta, iteration


# Sample usage
if __name__ == '__main__':
    # Load the system data
    with open("9 bus system.pkl", "rb") as file:
        system = pickle.load(file)
    # Perform the power flow
    v, theta, iteration = gauss_seidel(system)
    print("Voltage magnitudes:", v)
    print("Voltage phase angles:", theta)

File: Lecture 3/MATLAB/gauss_seidel.m
function [V,Theta,Iteration] = gauss_seidel(System)

% Extract the system data

Number_of_Buses = size(System.Buses, 1) ;
Bus_Type = System.Buses.Bus_Type ;
V_specified = System.Buses.V ;
Theta = System.Buses.Theta ;
P_load = System.Buses.P_load ;
Q_load = System.Buses.Q_load ;
P_gen = System.Buses.P_gen ;
Q_gen = System.Buses.Q_gen ;
Q_gen_min = System.Buses.Q_gen_min ;
Q_gen_max = System.Buses.Q_gen_max ;

% Create the bus-admittance matrix

Y_b = create_Yb(System) ;

% Initialize the calculation variables

P = P_gen - P_load ;
Q = Q_gen - Q_load ;
V = V_specified.*exp(sqrt(-1)*Theta) ;
V_old = zeros(size(V)) ;
Iteration = 0 ; 
Tolerance = 1e-6 ;

% Precompute reciprocal of diagonal elements of Y_b

Y_b_diag_inv = 1./diag(Y_b) ;

% Main loop

while any(abs(V - V_old) > Tolerance)

    % Update the iteration variables
    
    V_old = V ;
    Iteration = Iteration + 1 ; 

    % Apply the Gauss-Seidel power flow equations to each node except the slack node
    
    for i = 2 : Number_of_Buses

        % Calculate the sum of node current contributions

        S = Y_b(i,:)*V ;  
        S = S - Y_b(i,i)*V(i) ; 

        if Bus_Type(i) == 2

            % If the current node represents a PV node, determine the
            % reactive power injection
            
            Q(i) = - imag(conj(V(i))*(S+Y_b(i,i)*V(i))) ;
            Q_gen(i) = Q(i) + Q_load(i) ;

            % Check the generator reactive power limits
            
            if Q_gen(i) < Q_gen_min(i) || Q_gen(i) > Q_gen_max(i)
                
                % If reactive power limits are not satisfied, limit the
                % reactive power value and update the voltage magnitude and
                % phase angle

                Q_gen(i) = max(min(Q_gen(i),Q_gen_max(i)),Q_gen_min(i)) ;
                Q(i) = Q_gen(i) - Q_load(i) ;
                V(i) = Y_b_diag_inv(i)*((P(i)-sqrt(-1)*Q(i))/conj(V(i))-S) ;
            
            else

                % If reactive power limits are satisfied, update only the
                % voltage phase angle

                V(i) = Y_b_diag_inv(i)*((P(i)-sqrt(-1)*Q(i))/conj(V(i))-S) ;
                V(i) = abs(V_specified(i))*exp(sqrt(-1)*angle(V(i))) ;
            
            end
        
        else
            
            % If the current node represents a PQ node, update the complex
            % node voltage
            
            V(i) = Y_b_diag_inv(i)*((P(i)-sqrt(-1)*Q(i))/conj(V(i))-S) ;
        
        end
    
    end

end

% Evaluate the voltage magnitudes and phase angles

Theta = angle(V)*180/pi ;
V = abs(V) ;

end